apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/EphemeralTransaction.cs

```diff
@@ -95,22 +89,20 @@ async Task<IEnumerable<long>> Impl()
         /// <returns>The number of rows modified.</returns>
         public Task<int> ExecuteMutationsAsync(List<Mutation> mutations, CancellationToken cancellationToken, int timeoutSeconds)
         {
-            return ExecuteHelper.WithErrorTranslationAndProfiling(Impl, "EphemeralTransaction.ExecuteMutationsAsync", _connection.Logger);
+            return ExecuteInReadWriteTransactionWithErrorTranslationAndProfiling(Impl, cancellationToken, "EphemeralTransaction.ExecuteMutationsAsync");
 
-            async Task<int> Impl()
+            async Task<int> Impl(SpannerTransaction transaction, Action rollbackCheckpoint)
             {
-                using (var transaction = await _connection.BeginTransactionImplAsync(_transactionOptions, TransactionMode.ReadWrite, cancellationToken).ConfigureAwait(false))
-                {
-                    // Importantly, we need to set timeout on the transaction, because
-                    // ExecuteMutations on SpannerTransaction doesnt actually hit the network
-                    // until you commit or rollback.
-                    transaction.CommitTimeout = timeoutSeconds;
-                    int count = await ((ISpannerTransaction)transaction)
-                        .ExecuteMutationsAsync(mutations, cancellationToken, timeoutSeconds)
-                        .ConfigureAwait(false);
-                    await transaction.CommitAsync(cancellationToken).ConfigureAwait(false);
-                    return count;
-                }
+                // Importantly, we need to set timeout on the transaction, because
+                // ExecuteMutations on SpannerTransaction doesnt actually hit the network
+                // until you commit or rollback.
+                transaction.CommitTimeout = timeoutSeconds;
+                int count = await ((ISpannerTransaction)transaction)
+                    .ExecuteMutationsAsync(mutations, cancellationToken, timeoutSeconds)
+                    .ConfigureAwait(false);
+                rollbackCheckpoint();
```

Comment:
Should this be after the commit? What state is the transaction left in if the commit fails? And it doesn't hurt us to put this after the commit anyway (I think). And same for the others except dml which is of its own.

Reply:
No, it should be before the commit - because I believe if *commit* fails, then it's automatically rolled back and we don't need to call rollback. Although I suppose there could be a communication problem with means the reason the commit fails is that the server never gets the message.

I'll leave it for now, and we'll see what the Spanner team thinks we should do.